src/main/java/ui/UI.java

```diff
@@ -555,13 +556,19 @@ public HWND getMainWindowHandle() {
         return mainWindowHandle;
     }
 
-    public void showNotificationPane(Node graphic, String text, Action action) {
+    public void showNotificationPane(Node graphic, String text, Runnable runnable, Runnable undoRunnable) {
         Platform.runLater(() -> {
             hideNotificationPane();
+            notificationRunnable = Optional.of(runnable);
+            notificationUndoRunnable = Optional.of(undoRunnable);
             notificationPane.setGraphic(graphic);
             notificationPane.setText(text);
             notificationPane.getActions().clear();
-            notificationPane.getActions().add(action);
+            notificationPane.getActions().add(new Action("Undo",
+                    actionEvent -> Platform.runLater(() -> {
+                        notificationRunnable = Optional.of(undoRunnable);
```

Comment:
We should try to keep state local. Maybe a class in charge of controlling the notification panel?